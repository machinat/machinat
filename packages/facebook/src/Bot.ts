import { AgentSettingsAccessor } from '@sociably/core';
import Engine, { DispatchError } from '@sociably/core/engine';
import Queue from '@sociably/core/queue';
import Renderer from '@sociably/core/renderer';
import ModuleUtilitiesI from '@sociably/core/base/ModuleUtilities';
import { serviceProviderClass } from '@sociably/core/service';
import type {
  SociablyNode,
  SociablyBot,
  InitScopeFn,
  DispatchWrapper,
} from '@sociably/core';
import {
  MetaApiWorker,
  MetaApiJob,
  MetaApiResult,
  MetaApiDispatchResponse,
  MetaApiResponseBody,
} from '@sociably/meta-api';
import generalComponentDelegator from './components/general';
import { FACEBOOK, PATH_FEED, PATH_PHOTOS } from './constant';
import {
  ConfigsI,
  PlatformUtilitiesI,
  PageSettingsAccessorI,
} from './interface';
import FacebookPage from './Page';
import FacebookChat from './Chat';
import InteractTarget from './InteractTarget';
import {
  createChatJobs,
  createChatAttachmentJobs,
  createPostJobs,
  createInteractJobs,
} from './job';
import type {
  FacebookThread,
  FacebookComponent,
  FacebookSegmentValue,
  FacebookDispatchFrame,
  MessagingOptions,
  FacebookPageSettings,
} from './types';

type FacebookDispatchTarget = FacebookThread | FacebookPage;

type FacebookBotOptions = {
  appSecret?: string;
  graphApiVersion?: string;
  apiBatchRequestInterval?: number;
  pageSettingsAccessor: AgentSettingsAccessor<
    FacebookPage,
    FacebookPageSettings
  >;
  initScope?: InitScopeFn;
  dispatchWrapper?: DispatchWrapper<
    MetaApiJob,
    FacebookDispatchFrame,
    MetaApiResult
  >;
};

type UploadAttachmentResult = {
  attachmentId: string;
};

type PagePhotoResult = {
  photoId: string;
};

type PagePostResult = {
  postId: string;
  photos: PagePhotoResult[];
};

type CommentResult = {
  commentId: string;
  photo: null | PagePhotoResult;
};

type ApiCallOptions = {
  /** The page to make the API call */
  page: string | FacebookPage;
  /** HTTP method */
  method?: string;
  /** API request URL relative to https://graph.facebook.com/{version}/ */
  url: string;
  /** API request parameters */
  params?: Record<string, unknown>;
};

/**
 * FacebookBot render messages and make API call to Facebook platform.
 * @category Provider
 */
export class FacebookBot
  implements SociablyBot<FacebookThread, MetaApiJob, MetaApiResult>
{
  worker: MetaApiWorker;
  engine: Engine<
    FacebookDispatchTarget,
    FacebookSegmentValue,
    FacebookComponent<unknown>,
    MetaApiJob,
    MetaApiResult
  >;

  platform = FACEBOOK;

  constructor({
    appSecret,
    graphApiVersion = 'v11.0',
    apiBatchRequestInterval = 500,
    pageSettingsAccessor,
    initScope,
    dispatchWrapper,
  }: FacebookBotOptions) {
    const renderer = new Renderer<
      FacebookSegmentValue,
      FacebookComponent<unknown>
    >(FACEBOOK, generalComponentDelegator);

    const queue = new Queue<MetaApiJob, MetaApiResult>();
    const worker = new MetaApiWorker(
      pageSettingsAccessor,
      appSecret,
      graphApiVersion,
      apiBatchRequestInterval
    );

    this.engine = new Engine(
      FACEBOOK,
      renderer,
      queue,
      worker,
      initScope,
      dispatchWrapper
    );
  }

  async start(): Promise<void> {
    this.engine.start();
  }

  async stop(): Promise<void> {
    this.engine.stop();
  }

  async render(
    target: FacebookDispatchTarget,
    node: SociablyNode
  ): Promise<null | MetaApiDispatchResponse> {
    if (target instanceof FacebookChat) {
      return this.engine.render(target, node, createChatJobs());
    }
    if (target instanceof FacebookPage) {
      return this.engine.render(target, node, createPostJobs);
    }
    if (target instanceof InteractTarget) {
      return this.engine.render(target, node, createInteractJobs);
    }
    throw new TypeError('invalid rendering target');
  }

  /** Send messages or actions on a chat */
  async message(
    chat: FacebookChat,
    messages: SociablyNode,
    options?: MessagingOptions
  ): Promise<null | MetaApiDispatchResponse> {
    return this.engine.render(chat, messages, createChatJobs(options));
  }

  /** Upload a media chat attachment for later use */
  async uploadChatAttachment(
    /** The {@link FacebookPage} that owns the attachment */
    pageInput: string | FacebookPage,
    /** An {@link Image}, {@link Audio}, {@link Video} or {@link File} element to be uploaded */ node: SociablyNode
  ): Promise<null | UploadAttachmentResult> {
    const page =
      typeof pageInput === 'string' ? new FacebookPage(pageInput) : pageInput;
    const response = await this.engine.render(
      page,
      node,
      createChatAttachmentJobs
    );
    const result = response?.results[0].body;
    return result ? { attachmentId: result.attachment_id } : null;
  }

  /** Create a post or a photo on the page feed */
  async post(
    /** The {@link FacebookPage} to post */
    pageInput: string | FacebookPage,
    /** Text, a {@link PagePost} or a {@link PagePhoto} to post */
    node: SociablyNode
  ): Promise<null | PagePostResult> {
    const page =
      typeof pageInput === 'string' ? new FacebookPage(pageInput) : pageInput;
    const response = await this.engine.render(page, node, createPostJobs);
    if (!response) {
      return null;
    }

    const photos: PagePhotoResult[] = [];
    let postId: undefined | string;

    const { jobs, results } = response;
    for (const [i, { request }] of jobs.entries()) {
      const resultBody = results[i].body;

      if (request.url === PATH_FEED) {
        postId = resultBody.id;
      } else if (request.url === PATH_PHOTOS) {
        if (resultBody.post_id) {
          postId = resultBody.post_id;
        }
        photos.push({ photoId: resultBody.id });
      }
    }
    if (!postId) {
      return null;
    }
    return { postId, photos };
  }

  /** Comment or make reactions to a Facebook post or comment */
  async interact(
    /** The target thread to interact with */
    target: InteractTarget,
    /** {@link Comment} or {@link Reaction} */
    node: SociablyNode
  ): Promise<null | CommentResult[]> {
    const response = await this.engine.render(target, node, createInteractJobs);
    if (!response) {
      return null;
    }

    const results: CommentResult[] = [];
    let photo: null | PagePhotoResult = null;

    for (const [i, { request }] of response.jobs.entries()) {
      if (request.url === PATH_PHOTOS) {
        photo = { photoId: response.results[i].body.id };
      } else if (request.url === PATH_PHOTOS) {
        results.push({
          commentId: response.results[i].body.id,
          photo,
        });
        photo = null;
      }
    }
    return results;
  }

  async requestApi<ResBody extends MetaApiResponseBody>({
    page: pageInput,
    method = 'GET',
    url,
    params,
  }: ApiCallOptions): Promise<ResBody> {
    const page =
      typeof pageInput === 'string' ? new FacebookPage(pageInput) : pageInput;
    try {
      const { results } = await this.engine.dispatchJobs(page, [
        {
          channel: page,
          request: { method, url, params },
        },
      ]);

      return results[0].body as ResBody;
    } catch (err) {
      if (err instanceof DispatchError) {
        throw err.errors[0];
      }
      throw err;
    }
  }
}

const BotP = serviceProviderClass({
  lifetime: 'singleton',
  deps: [
    ConfigsI,
    PageSettingsAccessorI,
    { require: ModuleUtilitiesI, optional: true },
    { require: PlatformUtilitiesI, optional: true },
  ],
  factory: (
    { appSecret, apiBatchRequestInterval },
    pageSettingsAccessor,
    moduleUitils,
    platformUtils
  ) =>
    new FacebookBot({
      pageSettingsAccessor,
      appSecret,
      apiBatchRequestInterval,
      initScope: moduleUitils?.initScope,
      dispatchWrapper: platformUtils?.dispatchWrapper,
    }),
})(FacebookBot);

type BotP = FacebookBot;
export default BotP;
