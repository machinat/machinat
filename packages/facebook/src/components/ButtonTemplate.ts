import type { SociablyNode } from '@sociably/core';
import { formatNode } from '@sociably/core/utils';
import { makeUnitSegment } from '@sociably/core/renderer';
import type { UnitSegment } from '@sociably/core/renderer';
import makeFacebookComponent from '../utils/makeFacebookComponent.js';
import { PATH_MESSAGES } from '../constant.js';
import type { MessageValue, FacebookComponent } from '../types.js';

/**
 * @category Props
 */
export type ButtonTemplateProps = {
  /**
   * Textual node with content up to 640 characters. Text will appear above the
   * buttons.
   */
  children: SociablyNode;
  /** 1-3 button elements to append after the text. */
  buttons: SociablyNode;
  sharable?: boolean;
};

/**
 * The button template allows you to send a structured message that includes
 * text and buttons.
 * @category Component
 * @props {@link ButtonTemplateProps}
 * @guides Check official [doc](https://developers.facebook.com/docs/messenger-platform/send-messages/template/button)
 *   and [reference](https://developers.facebook.com/docs/messenger-platform/reference/templates/button).
 */
export const ButtonTemplate: FacebookComponent<
  ButtonTemplateProps,
  UnitSegment<MessageValue>
> = makeFacebookComponent(async function ButtonTemplate(node, path, render) {
  const { children, buttons, sharable } = node.props;
  const textSegments = await render(children, '.children');

  if (!textSegments) {
    throw new TypeError(`"children" prop should not be empty`);
  }
  for (const segment of textSegments) {
    if (segment.type !== 'text') {
      throw new TypeError(
        `non-textual node ${formatNode(
          segment.node
        )} received, only textual nodes allowed`
      );
    }
  }

  const buttonSegments = await render(buttons, '.buttons');
  const buttonValues = buttonSegments?.map((segment) => segment.value);

  return [
    makeUnitSegment(node, path, {
      type: 'message',
      apiPath: PATH_MESSAGES,
      params: {
        message: {
          attachment: {
            type: 'template',
            payload: {
              template_type: 'button',
              text: textSegments[0].value,
              sharable,
              buttons: buttonValues,
            },
          },
        },
      },
    }),
  ];
});
