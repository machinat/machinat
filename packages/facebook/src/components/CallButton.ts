import { makePartSegment } from '@sociably/core/renderer';
import type { PartSegment } from '@sociably/core/renderer';
import makeFacebookComponent from '../utils/makeFacebookComponent.js';
import type { FacebookComponent } from '../types.js';

/**
 * @category Props
 */
export type CallButtonProps = {
  /** Button title, 20 character limit. */
  title: string;
  /**
   * Format must have "+" prefix followed by the country code, area code and
   * local number.
   */
  number: string;
};

/**
 * The Call Button can be used to initiate a phone call. This button can be used
 * with the Button and Generic Templates.
 * @category Component
 * @props {@link CallButtonProps}
 * @guides Check official [doc](https://developers.facebook.com/docs/messenger-platform/send-messages/buttons#call)
 *   and [reference](https://developers.facebook.com/docs/messenger-platform/reference/buttons/call).
 */
export const CallButton: FacebookComponent<
  CallButtonProps,
  PartSegment<{}>
> = makeFacebookComponent(function CallButton(node, path) {
  const { title, number } = node.props;
  return [
    makePartSegment(node, path, {
      type: 'phone_number',
      title,
      number,
    }),
  ];
});
