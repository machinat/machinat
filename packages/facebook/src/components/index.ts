export * from './button';
export * from './handoverProtocol';
export * from './media';
export * from './quickReply';
export * from './senderAction';
export * from './template';
export * from './text';
export * from './expression';
