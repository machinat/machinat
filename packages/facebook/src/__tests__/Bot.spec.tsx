import querystring from 'querystring';
import nock from 'nock';
import moxy, { Moxy } from '@moxyjs/moxy';
import Sociably from '@sociably/core';
import _Renderer from '@sociably/core/renderer';
import Queue from '@sociably/core/queue';
import _Engine from '@sociably/core/engine';
import { MetaApiWorker as _Worker, MetaApiError } from '@sociably/meta-api';
import { FacebookBot } from '../Bot';
import { Image, Expression, TextReply } from '../components';
import FacebookChat from '../Chat';
import FacebookPage from '../Page';

const Renderer = _Renderer as Moxy<typeof _Renderer>;
const Engine = _Engine as Moxy<typeof _Engine>;
const Worker = _Worker as Moxy<typeof _Worker>;

nock.disableNetConnect();

jest.mock('@sociably/core/engine', () =>
  jest
    .requireActual('@moxyjs/moxy')
    .default(jest.requireActual('@sociably/core/engine'))
);

jest.mock('@sociably/core/renderer', () =>
  jest
    .requireActual('@moxyjs/moxy')
    .default(jest.requireActual('@sociably/core/renderer'))
);

jest.mock('@sociably/meta-api', () => {
  const module = jest.requireActual('@sociably/meta-api');
  return {
    ...module,
    MetaApiWorker: jest
      .requireActual('@moxyjs/moxy')
      .default(module.MetaApiWorker),
  };
});

const initScope = moxy(() => moxy());
const dispatchWrapper = moxy((x) => x);

const pageId = '1234567890';
const accessToken = '_ACCESS_TOKEN_';
const appSecret = '_APP_SECRET_';

const pageSettingsAccessor = {
  getChannelSettings: async () => ({ pageId, accessToken }),
  getChannelSettingsBatch: async () => [{ pageId, accessToken }],
  listAllChannelSettings: async () => [{ pageId, accessToken }],
};

const page = new FacebookPage(pageId);

const message = (
  <Expression quickReplies={<TextReply title="Hi!" payload="ðŸ‘‹" />}>
    Hello <b>World!</b>
    <Image url="https://sociably.io/greeting.png" />
  </Expression>
);

let graphApi;
const bodySpy = moxy(() => true);

beforeEach(() => {
  graphApi = nock('https://graph.facebook.com').post('/v11.0/', bodySpy);
  bodySpy.mock.clear();
  Engine.mock.clear();
  Renderer.mock.clear();
  Worker.mock.clear();
});

afterEach(() => {
  nock.cleanAll();
});

describe('.constructor(options)', () => {
  it('assemble core modules', () => {
    const bot = new FacebookBot({
      pageSettingsAccessor,
      initScope,
      dispatchWrapper,
    });

    expect(bot.engine).toBeInstanceOf(Engine);

    expect(Renderer).toHaveBeenCalledTimes(1);
    expect(Renderer).toHaveBeenCalledWith('facebook', expect.any(Function));

    expect(Engine).toHaveBeenCalledTimes(1);
    expect(Engine).toHaveBeenCalledWith(
      'facebook',
      expect.any(Renderer),
      expect.any(Queue),
      expect.any(Worker),
      initScope,
      dispatchWrapper
    );

    expect(Worker).toHaveBeenCalledTimes(1);
    expect(Worker).toHaveBeenCalledWith(
      pageSettingsAccessor,
      undefined,
      'v11.0',
      500
    );
  });

  it('pass options to worker', () => {
    expect(
      new FacebookBot({
        pageSettingsAccessor,
        initScope,
        dispatchWrapper,
        appSecret,
        graphApiVersion: 'v8.0',
        apiBatchRequestInterval: 0,
      })
    );

    expect(Worker).toHaveBeenCalledTimes(1);
    expect(Worker).toHaveBeenCalledWith(
      pageSettingsAccessor,
      '_APP_SECRET_',
      'v8.0',
      0
    );
  });
});

test('.start() and .stop() start/stop engine', () => {
  const bot = new FacebookBot({
    pageSettingsAccessor,
    initScope,
    dispatchWrapper,
    appSecret,
  });

  type MockEngine = Moxy<FacebookBot['engine']>;

  bot.start();
  expect((bot.engine as MockEngine).start).toHaveBeenCalledTimes(1);

  bot.stop();
  expect((bot.engine as MockEngine).stop).toHaveBeenCalledTimes(1);
});

describe('.message(thread, message, options)', () => {
  const bot = new FacebookBot({ pageSettingsAccessor, appSecret });

  let apiStatus;
  beforeEach(() => {
    const messageResult = {
      code: 200,
      body: JSON.stringify({ message_id: 'xxx', recipient_id: 'xxx' }),
    };
    apiStatus = graphApi.reply(200, [messageResult, messageResult]);
    bot.start();
  });

  afterEach(() => {
    bot.stop();
  });

  const chat = new FacebookChat('1234567890', { id: '9876543210' });

  it('resolves null if message is empty', async () => {
    const empties = [undefined, null, [], <></>];
    for (const empty of empties) {
      // eslint-disable-next-line no-await-in-loop
      await expect(bot.message(chat, empty)).resolves.toBe(null);
      expect(apiStatus.isDone()).toBe(false);
    }
  });

  it('send messages to me/messages api', async () => {
    const response = await bot.message(chat, message);
    expect(response).toMatchSnapshot();

    for (const result of response!.results) {
      expect(result).toEqual({
        code: 200,
        body: { message_id: 'xxx', recipient_id: 'xxx' },
      });
    }

    expect(bodySpy).toHaveBeenCalledTimes(1);
    const body = bodySpy.mock.calls[0].args[0];

    for (const request of JSON.parse(body.batch)) {
      expect(request.method).toBe('POST');
      expect(request.relative_url).toBe(
        'me/messages?access_token=_ACCESS_TOKEN_'
      );
    }

    expect(body).toMatchSnapshot();
    expect(JSON.parse(body.batch).map((req) => querystring.decode(req.body)))
      .toMatchInlineSnapshot(`
      Array [
        Object {
          "message": "{\\"text\\":\\"Hello World!\\"}",
          "recipient": "{\\"id\\":\\"9876543210\\"}",
        },
        Object {
          "message": "{\\"attachment\\":{\\"type\\":\\"image\\",\\"payload\\":{\\"url\\":\\"https://sociably.io/greeting.png\\"}},\\"quick_replies\\":[{\\"content_type\\":\\"text\\",\\"title\\":\\"Hi!\\",\\"payload\\":\\"ðŸ‘‹\\"}]}",
          "recipient": "{\\"id\\":\\"9876543210\\"}",
        },
      ]
    `);

    expect(apiStatus.isDone()).toBe(true);
  });

  test('message options', async () => {
    const response = await bot.message(chat, message, {
      messagingType: 'MESSAGE_TAG',
      tag: 'TRANSPORTATION_UPDATE',
      notificationType: 'SILENT_PUSH',
      personaId: 'billy17',
    });
    expect(response).toMatchSnapshot();

    for (const result of response!.results) {
      expect(result).toEqual({
        code: 200,
        body: { message_id: 'xxx', recipient_id: 'xxx' },
      });
    }

    expect(bodySpy).toHaveBeenCalledTimes(1);
    const body = bodySpy.mock.calls[0].args[0];

    expect(body).toMatchSnapshot();
    expect(JSON.parse(body.batch).map((req) => querystring.decode(req.body)))
      .toMatchInlineSnapshot(`
      Array [
        Object {
          "message": "{\\"text\\":\\"Hello World!\\"}",
          "messaging_type": "MESSAGE_TAG",
          "notification_type": "SILENT_PUSH",
          "persona_id": "billy17",
          "recipient": "{\\"id\\":\\"9876543210\\"}",
          "tag": "TRANSPORTATION_UPDATE",
        },
        Object {
          "message": "{\\"attachment\\":{\\"type\\":\\"image\\",\\"payload\\":{\\"url\\":\\"https://sociably.io/greeting.png\\"}},\\"quick_replies\\":[{\\"content_type\\":\\"text\\",\\"title\\":\\"Hi!\\",\\"payload\\":\\"ðŸ‘‹\\"}]}",
          "messaging_type": "MESSAGE_TAG",
          "notification_type": "SILENT_PUSH",
          "persona_id": "billy17",
          "recipient": "{\\"id\\":\\"9876543210\\"}",
          "tag": "TRANSPORTATION_UPDATE",
        },
      ]
    `);

    expect(apiStatus.isDone()).toBe(true);
  });
});

describe('.uploadChatAttachment(message)', () => {
  const bot = new FacebookBot({ pageSettingsAccessor, appSecret });

  beforeEach(() => {
    bot.start();
  });

  afterEach(() => {
    bot.stop();
  });

  it('resolves null if message is empty', async () => {
    const empties = [undefined, null, [], <></>];
    for (const empty of empties) {
      await expect(bot.uploadChatAttachment(page, empty)).resolves.toBe(null); // eslint-disable-line no-await-in-loop
    }
  });

  it('call message_attachments api', async () => {
    const apiStatus = graphApi.reply(200, [
      { code: 200, body: JSON.stringify({ attachment_id: 401759795 }) },
    ]);

    await expect(
      bot.uploadChatAttachment(
        page,
        <Image url="https://sociably.io/trollface.png" />
      )
    ).resolves.toEqual({ attachmentId: 401759795 });

    expect(bodySpy).toHaveBeenCalledTimes(1);
    const body = bodySpy.mock.calls[0].args[0];

    expect(body).toMatchSnapshot();
    const reqest = JSON.parse(body.batch)[0];
    expect(reqest.method).toBe('POST');
    expect(reqest.relative_url).toBe(
      'me/message_attachments?access_token=_ACCESS_TOKEN_'
    );
    expect(querystring.decode(reqest.body)).toMatchInlineSnapshot(`
      Object {
        "message": "{\\"attachment\\":{\\"type\\":\\"image\\",\\"payload\\":{\\"url\\":\\"https://sociably.io/trollface.png\\"}}}",
      }
    `);

    expect(apiStatus.isDone()).toBe(true);
  });
});

describe('.makeApiCall()', () => {
  it('call facebook graph api', async () => {
    const bot = new FacebookBot({ pageSettingsAccessor });
    bot.start();

    const apiCall = graphApi.reply(200, [{ code: 200, body: '{"foo":"bar"}' }]);

    await expect(
      bot.makeApiCall({
        page,
        method: 'POST',
        path: 'foo',
        params: { bar: 'baz' },
      })
    ).resolves.toEqual({
      foo: 'bar',
    });

    expect(apiCall.isDone()).toBe(true);
  });

  it('throw MetaApiError if api call fail', async () => {
    const bot = new FacebookBot({ pageSettingsAccessor });
    bot.start();

    const apiCall = graphApi.reply(200, [
      {
        code: 400,
        body: JSON.stringify({
          error: {
            message: 'bad',
            type: 'devil',
            code: 444,
            error_subcode: 666,
            fbtrace_id: 'xxxx',
          },
        }),
      },
    ]);

    try {
      await bot.makeApiCall({
        page,
        method: 'POST',
        path: 'foo',
        params: { bar: 'baz' },
      });
      expect('should not be here').toBeFalsy();
    } catch (err) {
      expect(err).toBeInstanceOf(MetaApiError);
      expect(err.code).toBe(444);
      expect(err.message).toBe('bad');
    }

    expect(apiCall.isDone()).toBe(true);
  });
});
