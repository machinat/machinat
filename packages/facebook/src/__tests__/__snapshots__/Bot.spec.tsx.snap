// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.message(thread, message, options) message options 1`] = `
Object {
  "jobs": Array [
    Object {
      "channel": FacebookPage {
        "id": "1234567890",
        "platform": "facebook",
      },
      "file": undefined,
      "key": "fb.1234567890.9876543210",
      "request": Object {
        "method": "POST",
        "params": Object {
          "message": Object {
            "text": "Hello World!",
          },
          "messaging_type": "MESSAGE_TAG",
          "notification_type": "SILENT_PUSH",
          "persona_id": "billy17",
          "recipient": Object {
            "id": "9876543210",
          },
          "tag": "TRANSPORTATION_UPDATE",
        },
        "url": "me/messages",
      },
    },
    Object {
      "channel": FacebookPage {
        "id": "1234567890",
        "platform": "facebook",
      },
      "file": undefined,
      "key": "fb.1234567890.9876543210",
      "request": Object {
        "method": "POST",
        "params": Object {
          "message": Object {
            "attachment": Object {
              "payload": Object {
                "attachment_id": undefined,
                "is_reusable": undefined,
                "url": "https://sociably.io/greeting.png",
              },
              "type": "image",
            },
            "metadata": undefined,
            "quick_replies": Array [
              Object {
                "content_type": "text",
                "image_url": undefined,
                "payload": "ðŸ‘‹",
                "title": "Hi!",
              },
            ],
          },
          "messaging_type": "MESSAGE_TAG",
          "notification_type": "SILENT_PUSH",
          "persona_id": "billy17",
          "recipient": Object {
            "id": "9876543210",
          },
          "tag": "TRANSPORTATION_UPDATE",
        },
        "url": "me/messages",
      },
    },
  ],
  "results": Array [
    Object {
      "body": Object {
        "message_id": "xxx",
        "recipient_id": "xxx",
      },
      "code": 200,
    },
    Object {
      "body": Object {
        "message_id": "xxx",
        "recipient_id": "xxx",
      },
      "code": 200,
    },
  ],
  "tasks": Array [
    Object {
      "payload": Array [
        Object {
          "channel": FacebookPage {
            "id": "1234567890",
            "platform": "facebook",
          },
          "file": undefined,
          "key": "fb.1234567890.9876543210",
          "request": Object {
            "method": "POST",
            "params": Object {
              "message": Object {
                "text": "Hello World!",
              },
              "messaging_type": "MESSAGE_TAG",
              "notification_type": "SILENT_PUSH",
              "persona_id": "billy17",
              "recipient": Object {
                "id": "9876543210",
              },
              "tag": "TRANSPORTATION_UPDATE",
            },
            "url": "me/messages",
          },
        },
        Object {
          "channel": FacebookPage {
            "id": "1234567890",
            "platform": "facebook",
          },
          "file": undefined,
          "key": "fb.1234567890.9876543210",
          "request": Object {
            "method": "POST",
            "params": Object {
              "message": Object {
                "attachment": Object {
                  "payload": Object {
                    "attachment_id": undefined,
                    "is_reusable": undefined,
                    "url": "https://sociably.io/greeting.png",
                  },
                  "type": "image",
                },
                "metadata": undefined,
                "quick_replies": Array [
                  Object {
                    "content_type": "text",
                    "image_url": undefined,
                    "payload": "ðŸ‘‹",
                    "title": "Hi!",
                  },
                ],
              },
              "messaging_type": "MESSAGE_TAG",
              "notification_type": "SILENT_PUSH",
              "persona_id": "billy17",
              "recipient": Object {
                "id": "9876543210",
              },
              "tag": "TRANSPORTATION_UPDATE",
            },
            "url": "me/messages",
          },
        },
      ],
      "type": "dispatch",
    },
  ],
}
`;

exports[`.message(thread, message, options) message options 2`] = `
Object {
  "access_token": "_ACCESS_TOKEN_",
  "appsecret_proof": "932e1d758c8379099e1b7f9e75e1abf41ab496760d64ddb05e3d21979d13c31f",
  "batch": "[{\\"method\\":\\"POST\\",\\"relative_url\\":\\"me/messages?access_token=_ACCESS_TOKEN_\\",\\"body\\":\\"message=%7B%22text%22%3A%22Hello%20World!%22%7D&messaging_type=MESSAGE_TAG&tag=TRANSPORTATION_UPDATE&notification_type=SILENT_PUSH&persona_id=billy17&recipient=%7B%22id%22%3A%229876543210%22%7D\\",\\"name\\":\\"fb.1234567890.9876543210-1\\",\\"omit_response_on_success\\":false},{\\"method\\":\\"POST\\",\\"relative_url\\":\\"me/messages?access_token=_ACCESS_TOKEN_\\",\\"body\\":\\"message=%7B%22attachment%22%3A%7B%22type%22%3A%22image%22%2C%22payload%22%3A%7B%22url%22%3A%22https%3A%2F%2Fsociably.io%2Fgreeting.png%22%7D%7D%2C%22quick_replies%22%3A%5B%7B%22content_type%22%3A%22text%22%2C%22title%22%3A%22Hi!%22%2C%22payload%22%3A%22%F0%9F%91%8B%22%7D%5D%7D&messaging_type=MESSAGE_TAG&tag=TRANSPORTATION_UPDATE&notification_type=SILENT_PUSH&persona_id=billy17&recipient=%7B%22id%22%3A%229876543210%22%7D\\",\\"depends_on\\":\\"fb.1234567890.9876543210-1\\",\\"name\\":\\"fb.1234567890.9876543210-2\\",\\"omit_response_on_success\\":false}]",
}
`;

exports[`.message(thread, message, options) send messages to me/messages api 1`] = `
Object {
  "jobs": Array [
    Object {
      "channel": FacebookPage {
        "id": "1234567890",
        "platform": "facebook",
      },
      "file": undefined,
      "key": "fb.1234567890.9876543210",
      "request": Object {
        "method": "POST",
        "params": Object {
          "message": Object {
            "text": "Hello World!",
          },
          "messaging_type": undefined,
          "notification_type": undefined,
          "persona_id": undefined,
          "recipient": Object {
            "id": "9876543210",
          },
          "tag": undefined,
        },
        "url": "me/messages",
      },
    },
    Object {
      "channel": FacebookPage {
        "id": "1234567890",
        "platform": "facebook",
      },
      "file": undefined,
      "key": "fb.1234567890.9876543210",
      "request": Object {
        "method": "POST",
        "params": Object {
          "message": Object {
            "attachment": Object {
              "payload": Object {
                "attachment_id": undefined,
                "is_reusable": undefined,
                "url": "https://sociably.io/greeting.png",
              },
              "type": "image",
            },
            "metadata": undefined,
            "quick_replies": Array [
              Object {
                "content_type": "text",
                "image_url": undefined,
                "payload": "ðŸ‘‹",
                "title": "Hi!",
              },
            ],
          },
          "messaging_type": undefined,
          "notification_type": undefined,
          "persona_id": undefined,
          "recipient": Object {
            "id": "9876543210",
          },
          "tag": undefined,
        },
        "url": "me/messages",
      },
    },
  ],
  "results": Array [
    Object {
      "body": Object {
        "message_id": "xxx",
        "recipient_id": "xxx",
      },
      "code": 200,
    },
    Object {
      "body": Object {
        "message_id": "xxx",
        "recipient_id": "xxx",
      },
      "code": 200,
    },
  ],
  "tasks": Array [
    Object {
      "payload": Array [
        Object {
          "channel": FacebookPage {
            "id": "1234567890",
            "platform": "facebook",
          },
          "file": undefined,
          "key": "fb.1234567890.9876543210",
          "request": Object {
            "method": "POST",
            "params": Object {
              "message": Object {
                "text": "Hello World!",
              },
              "messaging_type": undefined,
              "notification_type": undefined,
              "persona_id": undefined,
              "recipient": Object {
                "id": "9876543210",
              },
              "tag": undefined,
            },
            "url": "me/messages",
          },
        },
        Object {
          "channel": FacebookPage {
            "id": "1234567890",
            "platform": "facebook",
          },
          "file": undefined,
          "key": "fb.1234567890.9876543210",
          "request": Object {
            "method": "POST",
            "params": Object {
              "message": Object {
                "attachment": Object {
                  "payload": Object {
                    "attachment_id": undefined,
                    "is_reusable": undefined,
                    "url": "https://sociably.io/greeting.png",
                  },
                  "type": "image",
                },
                "metadata": undefined,
                "quick_replies": Array [
                  Object {
                    "content_type": "text",
                    "image_url": undefined,
                    "payload": "ðŸ‘‹",
                    "title": "Hi!",
                  },
                ],
              },
              "messaging_type": undefined,
              "notification_type": undefined,
              "persona_id": undefined,
              "recipient": Object {
                "id": "9876543210",
              },
              "tag": undefined,
            },
            "url": "me/messages",
          },
        },
      ],
      "type": "dispatch",
    },
  ],
}
`;

exports[`.message(thread, message, options) send messages to me/messages api 2`] = `
Object {
  "access_token": "_ACCESS_TOKEN_",
  "appsecret_proof": "932e1d758c8379099e1b7f9e75e1abf41ab496760d64ddb05e3d21979d13c31f",
  "batch": "[{\\"method\\":\\"POST\\",\\"relative_url\\":\\"me/messages?access_token=_ACCESS_TOKEN_\\",\\"body\\":\\"message=%7B%22text%22%3A%22Hello%20World!%22%7D&recipient=%7B%22id%22%3A%229876543210%22%7D\\",\\"name\\":\\"fb.1234567890.9876543210-1\\",\\"omit_response_on_success\\":false},{\\"method\\":\\"POST\\",\\"relative_url\\":\\"me/messages?access_token=_ACCESS_TOKEN_\\",\\"body\\":\\"message=%7B%22attachment%22%3A%7B%22type%22%3A%22image%22%2C%22payload%22%3A%7B%22url%22%3A%22https%3A%2F%2Fsociably.io%2Fgreeting.png%22%7D%7D%2C%22quick_replies%22%3A%5B%7B%22content_type%22%3A%22text%22%2C%22title%22%3A%22Hi!%22%2C%22payload%22%3A%22%F0%9F%91%8B%22%7D%5D%7D&recipient=%7B%22id%22%3A%229876543210%22%7D\\",\\"depends_on\\":\\"fb.1234567890.9876543210-1\\",\\"name\\":\\"fb.1234567890.9876543210-2\\",\\"omit_response_on_success\\":false}]",
}
`;

exports[`.uploadChatAttachment(page, message) call message_attachments api 1`] = `
Object {
  "access_token": "_ACCESS_TOKEN_",
  "appsecret_proof": "932e1d758c8379099e1b7f9e75e1abf41ab496760d64ddb05e3d21979d13c31f",
  "batch": "[{\\"method\\":\\"POST\\",\\"relative_url\\":\\"me/message_attachments?access_token=_ACCESS_TOKEN_\\",\\"body\\":\\"message=%7B%22attachment%22%3A%7B%22type%22%3A%22image%22%2C%22payload%22%3A%7B%22url%22%3A%22https%3A%2F%2Fsociably.io%2Ftrollface.png%22%7D%7D%7D\\",\\"omit_response_on_success\\":false}]",
}
`;
